File: backend/models/devices.go
package models

import "time"

type Device struct {
	ID             uint       `json:"id" gorm:"primaryKey"`
	IMEI           string     `json:"imei" gorm:"type:varchar(255);uniqueIndex;not null"`
	Name           string     `json:"name" gorm:"type:varchar(255)"` // User editable device name
	DeviceGroupID  uint       `json:"device_group_id" gorm:"not null"`
	DeviceGroup    string     `json:"device_group" gorm:"type:varchar(255);not null"`
	CountrySiteID  uint       `json:"country_site_id" gorm:"not null"`
	CountrySite    string     `json:"country_site" gorm:"type:varchar(255);not null"`
	DeviceType     string     `json:"device_type" gorm:"type:varchar(50);default:'android'"`
	Manufacturer   string     `json:"manufacturer" gorm:"type:varchar(255)"`
	Model          string     `json:"model" gorm:"type:varchar(255)"`
	AndroidVersion string     `json:"android_version" gorm:"type:varchar(50)"`
	BatteryLevel   int        `json:"battery_level"`
	BatteryStatus  string     `json:"battery_status" gorm:"type:varchar(50)"`
	SignalStrength int        `json:"signal_strength"`
	SignalDBM      int        `json:"signal_dbm"`
	NetworkType    string     `json:"network_type" gorm:"type:varchar(50)"`
	Latitude       float64    `json:"latitude"`
	Longitude      float64    `json:"longitude"`
	IsActive       bool       `json:"is_active" gorm:"default:true"`
	IsOnline       bool       `json:"is_online" gorm:"default:false"`
	LastSeen       *time.Time `json:"last_seen"`

	// Maintenance mode fields
	MaintenanceMode      bool       `json:"maintenance_mode" gorm:"default:false"`
	MaintenanceReason    string     `json:"maintenance_reason" gorm:"type:varchar(500)"`
	MaintenanceStartedAt *time.Time `json:"maintenance_started_at"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type DeviceSimCard struct {
	ID                  uint      `json:"id" gorm:"primaryKey"`
	DeviceIMEI          string    `json:"device_imei" gorm:"type:varchar(255);not null;index"`
	SlotIndex           int       `json:"slot_index" gorm:"not null"`
	SubscriptionID      int       `json:"subscription_id" gorm:"not null"`
	DisplayName         string    `json:"display_name" gorm:"type:varchar(255)"`
	CarrierName         string    `json:"carrier_name" gorm:"type:varchar(255)"`
	CountryISO          string    `json:"country_iso" gorm:"type:varchar(10)"`
	PhoneNumber         string    `json:"phone_number" gorm:"type:varchar(50)"`
	NetworkMCC          string    `json:"network_mcc" gorm:"type:varchar(10)"`
	NetworkMNC          string    `json:"network_mnc" gorm:"type:varchar(10)"`
	SimMCC              string    `json:"sim_mcc" gorm:"type:varchar(10)"`
	SimMNC              string    `json:"sim_mnc" gorm:"type:varchar(10)"`
	NetworkOperatorName string    `json:"network_operator_name" gorm:"type:varchar(255)"`
	SimOperatorName     string    `json:"sim_operator_name" gorm:"type:varchar(255)"`
	Roaming             bool      `json:"roaming" gorm:"default:false"`
	IsActive            bool      `json:"is_active" gorm:"default:true"`
	IMEI                string    `json:"imei" gorm:"type:varchar(255)"`
	IMSI                string    `json:"imsi" gorm:"type:varchar(255)"`
	ICCID               string    `json:"iccid" gorm:"type:varchar(255)"`
	SignalStrength      int       `json:"signal_strength"`
	SignalDBM           int       `json:"signal_dbm"`
	SignalType          string    `json:"signal_type" gorm:"type:varchar(50)"`
	RSRP                int       `json:"rsrp"`
	RSRQ                int       `json:"rsrq"`
	RSSNR               int       `json:"rssnr"`
	CQI                 int       `json:"cqi"`
	NetworkType         string    `json:"network_type" gorm:"type:varchar(50)"`
	TotalDelivered      int       `json:"total_delivered" gorm:"default:0"`
	TotalSent           int       `json:"total_sent" gorm:"default:0"`
	TotalWaiting        int       `json:"total_waiting" gorm:"default:0"`
	MainBalance         float64   `json:"main_balance" gorm:"default:0"`
	SmsBalance          int       `json:"sms_balance" gorm:"default:0"`
	SmsLimit            int       `json:"sms_limit" gorm:"default:0"`
	DeviceName          string    `json:"device_name" gorm:"type:varchar(255)"`
	CountrySite         string    `json:"country_site" gorm:"type:varchar(255)"`
	DeviceGroupName     string    `json:"device_group_name" gorm:"type:varchar(255)"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
}

File: frontend/src/services/devices.ts
import { apiClient } from './api-client';

export interface Device {
  id: number;
  imei: string;
  name: string;
  device_group_id: number;
  device_group: string;
  country_site_id: number;
  country_site: string;
  device_type: string;
  manufacturer: string;
  model: string;
  android_version: string;
  battery_level: number;
  battery_status: string;
  signal_strength: number;
  signal_dbm: number;
  network_type: string;
  latitude: number;
  longitude: number;
  is_active: boolean;
  is_online: boolean;
  maintenance_mode: boolean;
  maintenance_reason: string;
  maintenance_started_at: string;
  last_seen: string;
  created_at: string;
  updated_at: string;
  alarms?: any[];
  // SMS Limit tracking fields
  sim1_daily_sms_used?: number;
  sim1_monthly_sms_used?: number;
  sim1_daily_limit_reset_at?: string;
  sim1_monthly_limit_reset_at?: string;
  sim2_daily_sms_used?: number;
  sim2_monthly_sms_used?: number;
  sim2_daily_limit_reset_at?: string;
  sim2_monthly_limit_reset_at?: string;
  
  // SMS Limit fields from Device Group (for display purposes)
  sim1_daily_sms_limit?: number;
  sim1_monthly_sms_limit?: number;
  sim2_daily_sms_limit?: number;
  sim2_monthly_sms_limit?: number;
  enable_sms_limits?: boolean;
}

export interface AlarmLog {
  id: number;
  device_id: string;
  device_name: string;
  device_group: string;
  country_site: string;
  alarm_type: string;
  message: string;
  severity: string;
  status: string;
  battery_level: number;
  battery_status: string;
  signal_strength: number;
  signal_dbm: number;
  network_type: string;
  created_at: string;
  updated_at: string;
}

export interface DeviceStats {
  total: number;
  active: number;
  inactive: number;
  online: number;
  offline: number;
  maintenance: number;
  ready: number;
  alarm: number;
}

export interface DevicesResponse {
  data: Device[];
  total: number;
  page: number;
  limit: number;
}

export interface AlarmLogsResponse {
  data: AlarmLog[];
  total: number;
  page: number;
  limit: number;
}

export const deviceService = {
  // Get all devices with pagination
  getDevices: async (params?: {
    page?: number;
    limit?: number;
    search?: string;
    country_site?: string;
    device_group?: string;
    status?: string;
    online?: string;
    maintenance?: string;
  }): Promise<DevicesResponse> => {
    const searchParams = new URLSearchParams();
    if (params?.page) searchParams.append('page', params.page.toString());
    if (params?.limit) searchParams.append('limit', params.limit.toString());
    if (params?.search) searchParams.append('search', params.search);
    if (params?.country_site) searchParams.append('country_site', params.country_site);
    if (params?.device_group) searchParams.append('device_group', params.device_group);
    if (params?.status) searchParams.append('status', params.status);
    if (params?.online) searchParams.append('online', params.online);
    if (params?.maintenance) searchParams.append('maintenance', params.maintenance);

    return apiClient.get<DevicesResponse>(`/devices?${searchParams.toString()}`);
  },

  // Get device by IMEI
  getDevice: async (imei: string): Promise<Device> => {
    return apiClient.get<Device>(`/devices/${imei}`);
  },

  // Get device statistics
  getDeviceStats: async (): Promise<DeviceStats> => {
    const response = await apiClient.get<{ success: boolean; data: DeviceStats }>('/devices/stats');
    return response.data;
  },

  // Delete devices
  deleteDevices: async (imeis: string[]): Promise<{ success: boolean }> => {
    const searchParams = new URLSearchParams();
    imeis.forEach(imei => searchParams.append('imeis', imei));
    return apiClient.delete<{ success: boolean }>(`/devices?${searchParams.toString()}`);
  },

  // Toggle device active status
  toggleDeviceActive: async (imeis: string[], active: boolean): Promise<{ success: boolean }> => {
    return apiClient.put<{ success: boolean }>('/devices/toggle-active', {
      imeis,
      active
    });
  },

  // Enter maintenance mode
  enterMaintenanceMode: async (imeis: string[], reason: string): Promise<{ success: boolean }> => {
    return apiClient.post<{ success: boolean }>('/devices/maintenance/enter', {
      imeis,
      reason
    });
  },

  // Exit maintenance mode
  exitMaintenanceMode: async (imeis: string[]): Promise<{ success: boolean }> => {
    return apiClient.post<{ success: boolean }>('/devices/maintenance/exit', {
      imeis
    });
  },

  // Rename device
  renameDevice: async (imei: string, name: string): Promise<{ success: boolean }> => {
    return apiClient.put<{ success: boolean }>(`/devices/${imei}/rename`, {
      name
    });
  },

  // Get alarm logs
  getAlarmLogs: async (params?: {
    page?: number;
    limit?: number;
    search?: string;
    device_id?: string;
    alarm_type?: string;
    severity?: string;
    status?: string;
    start_date?: string;
    end_date?: string;
  }): Promise<AlarmLogsResponse> => {
    const searchParams = new URLSearchParams();
    if (params?.page) searchParams.append('page', params.page.toString());
    if (params?.limit) searchParams.append('limit', params.limit.toString());
    else searchParams.append('limit', '50'); // Default to 50 records
    if (params?.search) searchParams.append('search', params.search);
    if (params?.device_id) searchParams.append('device_id', params.device_id);
    if (params?.alarm_type) searchParams.append('alarm_type', params.alarm_type);
    if (params?.severity) searchParams.append('severity', params.severity);
    if (params?.status) searchParams.append('status', params.status);
    if (params?.start_date) searchParams.append('start_date', params.start_date);
    if (params?.end_date) searchParams.append('end_date', params.end_date);

    return apiClient.get<AlarmLogsResponse>(`/alarm-logs?${searchParams.toString()}`);
  },

  // Clear alarm logs
  clearAlarmLogs: async (): Promise<{ success: boolean }> => {
    return apiClient.delete<{ success: boolean }>('/alarm-logs');
  },

  // Get SMS limit status for a device
  getSmsLimitStatus: async (imei: string, simSlot: number): Promise<any> => {
    return apiClient.get(`/devices/${imei}/sms-limits/status?sim_slot=${simSlot}`);
  },

  // Reset SMS limits for a device
  resetSmsLimits: async (imei: string, simSlot: number, resetType: string): Promise<{ success: boolean }> => {
    return apiClient.post(`/devices/${imei}/sms-limits/reset`, {
      sim_slot: simSlot,
      reset_type: resetType
    });
  },

  // Bulk update SMS limit data for all devices
  bulkUpdateSmsLimitData: async (): Promise<{ 
    success: boolean; 
    data?: {
      updated_devices: number;
      total_devices: number;
      device_groups_processed: number;
    };
  }> => {
    return apiClient.post('/devices/bulk-update-sms-limits');
  }
}; 